src/phoenix/core/model_schema.py

```diff
@@ -0,0 +1,1067 @@
+        with dummy dimensions for ones omitted by user. The dummy dimensions
+        have randomly generated names that can change for each iteration, but
+        currently there's no need to iterate more than once."""
+        for spec, role, data_type in chain(
+            (
+                (self.prediction_id, PREDICTION_ID, DISCRETE),
+                (self.timestamp, TIME, CONTINUOUS),
+                (self.prediction_label, PREDICTION_LABEL, DISCRETE),
+                (self.prediction_score, PREDICTION_SCORE, CONTINUOUS),
+                (self.actual_label, ACTUAL_LABEL, DISCRETE),
+                (self.actual_score, ACTUAL_SCORE, CONTINUOUS),
+            ),
+            zip(self.features, repeat(FEATURE), repeat(UNKNOWN)),
+            zip(self.tags, repeat(TAG), repeat(UNKNOWN)),
+        ):
+            if not isinstance(spec, CompositeDimensionSpec):
+                spec = _ensure_str(spec)
+            assert isinstance(role, DimensionRole)  # for mypy
+            if isinstance(spec, str):
+                yield ScalarDimension(spec, role=role, data_type=data_type)
+            elif isinstance(spec, Embedding):
+                yield EmbeddingDimension.from_(spec, role=role, data_type=data_type)
+            else:
+                raise TypeError("%s has unrecognized type: %s" % (role, type(spec)))
+
+    def __call__(
+        self,
+        *dataframes: Union[pd.DataFrame, Tuple[Name, pd.DataFrame]],
+        **kwargs: Any,
+    ) -> Model:
+        """Dimensions are the "baton" that Schema hands over to Model."""
+        _raise_if_too_many_dataframes(len(dataframes))
+        return Model(iter(self._dimensions), dataframes, **kwargs)
+
+    def to_json(self, **kwargs: Any) -> str:
+        return json.dumps(_jsonify(self), **kwargs)
+
+    @classmethod
+    def from_json(cls, json_string: str) -> Any:
+        json_data = json.loads(json_string)
+        return cls(**_objectify(json_data))
+
+
+def _agg_min_max(series: "pd.Series[Any]") -> "pd.Series[Any]":
+    return series.agg(["min", "max"])
+
+
+def _get_omitted_column_names(
+    dimensions: Iterable[Dimension],
+    dataframes: Iterable[pd.DataFrame],
+) -> Iterator[str]:
+    dataframe_columns = chain.from_iterable(df.columns for df in dataframes)
+    schema_columns = chain.from_iterable(dimensions)
+    yield from set(dataframe_columns) - set(schema_columns)
+
+
+def _group_names_by_dim_role(
+    dimensions: Iterable[Dimension],
+) -> Iterator[Tuple[DimensionRole, List[str]]]:
+    for role, dims in groupby(sorted(dimensions, key=_dim_role), key=_dim_role):
+        yield role, list(map(str, dims))
+
+
+def _guess_data_type(series: Iterable["pd.Series[Any]"]) -> DataType:
+    for s in series:
+        if is_numeric_dtype(s) or is_datetime64_any_dtype(s):
+            return CONTINUOUS
+    return DISCRETE
+
+
+_H = TypeVar("_H", bound=Hashable)
+
+
+def _dedup(it: Iterable[_H]) -> Iterator[_H]:
```

Comment:
```suggestion
def _dedup(iterable: Iterable[_H]) -> Iterator[_H]:
```
For function calls can you try to make the call signature human readable? Best to emulate languages that do this well like swift: https://www.swift.org/documentation/api-design-guidelines/